### ISSUE TITLE
Pickling a JAX array does not preserve its original device

### ISSUE THREAD

**Issue by ayaka14732** [google/jax#12253]
### What I hope to achieve

The JAX arrays are created in a subprocess, evoked by `multiprocessing.map`, on CPU. I hope that when the arrays are sent to the main process, they are still on the same device (i.e. CPU). However, after they are sent to the main process, they are placed on the default device (i.e. GPU) instead.

```python
import jax
import jax.numpy as np
import multiprocessing

def f():
    jax.config.update('jax_platforms', 'cpu')
    a = np.zeros((2,))
    print(a.device())  # TFRT_CPU_0
    return a

def main():
    ctx = multiprocessing.get_context('spawn')
    with ctx.Pool(1) as p:
        a = p.starmap(f, ((),))[0]
    print(a.device())  # gpu:0, expected TFRT_CPU_0

if __name__ == '__main__':
    main()
```

### This also applies to the underlying pickle module

```python
import jax
import jax.numpy as np
import pickle

jax.config.update('jax_platforms', 'cpu')

a = np.zeros((2,))
print(a.device())  # TFRT_CPU_0

with open('1.dat', 'wb') as f:
    pickle.dump(a, f)
```

```python
import pickle

with open('1.dat', 'rb') as f:
    a = pickle.load(f)
print(a.device())  # gpu:0, expected TFRT_CPU_0
```

### What jax/jaxlib version are you using?

jax 0.3.17, jaxlib 0.3.15+cuda11.cudnn82

### Which accelerator(s) are you using?

GPU

### Additional System Info

Python 3.10.6, Arch Linux x86_64

**Comment by jakevdp** [comment]
This is expected behavior; see the description from the CHANGELOG entry when `pickle` support was added: https://github.com/google/jax/blob/main/CHANGELOG.md#jax-0314-june-27-2022

The issue is that pickling and unpickling need not happen in the same environment, and so attempting to preserve the original device can be problematic; see https://github.com/google/jax/pull/10659 for further discussion, and let us know if you have any suggestions.

**Comment by ayaka14732** [comment]
@jakevdp Thank you! I agree that attempting to preserve the original device can be problematic. My suggestion is that we should give priority to implementing a `with` block to specify the target device. For example: in PyTorch we can do:

```python
with torch.cuda.device(2):  # specify default device
   ...  # create array on the default device
```

I remember that this issue was discussed in https://github.com/google/jax/issues/8879

**Comment by mattjj** [comment]
https://github.com/google/jax/pull/9118 may have fixed this, but I'm not sure. Checking with @skye ...

**Comment by skye** [comment]
Yes, I believe doing something like `with jax.default_device(jax.devices("cpu")[0])` should work. Please let me know if it doesn't, I haven't tried it with pickle.

**Comment by jakevdp** [comment]
Unpickling of an array [uses `jax.device_put`](https://github.com/google/jax/blob/3243e23aa528db390fdece3fa32517c28a50318b/jax/_src/device_array.py#L326) with no device argument, so I believe the default default device context manager should do the right thing.

**Comment by mattjj** [comment]
To put a fine point on it, the analogue of code like this:

```python
with torch.cuda.device(2):  # specify default device
   ...  # create array on the default device
```

Is something like this:
```python
devices = jax.devices()
with jax.default_device(devices[2]):
  ... # create or load (e.g. unpickle) array on default device
```

I think this issue is resolved, but please reopen if not!

**Comment by ayaka14732** [comment]
@mattjj Thank you for the reply, but this issue is not solved. `jax.default_device` works when creating the array, but does not work when loading the array from pickle:

```python
import jax
import jax.numpy as np
import multiprocessing

def f():
    jax.config.update('jax_platforms', 'cpu')
    a = np.zeros((2,))
    print(a.device())  # TFRT_CPU_0
    return a

def main():
    device_cpu = jax.devices('cpu')[0]
    with jax.default_device(device_cpu):
        a = np.zeros((2,))
        print(a.device())  # TFRT_CPU_0

        ctx = multiprocessing.get_context('spawn')
        with ctx.Pool(1) as p:
            a = p.starmap(f, ((),))[0]
        print(a.device())  # gpu:0, expected TFRT_CPU_0

if __name__ == '__main__':
    main()
```

**Comment by ayaka14732** [comment]
I found the cause of the problem

> Unpickling of an array [uses jax.device_put](https://github.com/google/jax/blob/3243e23aa528db390fdece3fa32517c28a50318b/jax/_src/device_array.py#L326) with no device argument, so I believe the default default device context manager should do the right thing.

But actually this does not work as expected:

```python
import jax
import numpy as onp

device_cpu = jax.devices('cpu')[0]
with jax.default_device(device_cpu):
    a = onp.array([1., 2.])

    b = jax.device_put(a)
    print(b.device())  # gpu:0, not TFRT_CPU_0
```

**Comment by ayaka14732** [comment]
@mattjj Please reopen this

**Comment by jakevdp** [comment]
Reopening and assigning to @skye – it looks like `device_put` does not respect the default device context. Is this intended?

**Comment by mattjj** [comment]
I think that is *not* intended.

(@ayaka14732 thanks for following up and for the pings. Sorry I didn't notice your replies until now!)

**Comment by meereeum** [comment]
Are there any updates on this? I'm wondering if I may be running into a related problem..

**TL;DR** recently updated jax, and code that used to run fine---using multiprocessing (forkserver) + jax (with default device / all arrays placed on CPU)---is now crashing due to memory overload on GPU. Monitoring the GPU, I can see that it starts by allocating 10Mb per process, which over time balloons to 100-400+, and ultimately crashes.

Thinking this might be caused by pickle (underlying multiprocessing) not playing well with device placement
